Add SegmentAllocationQueue to batch allocation actions

kfaraz opened the pull request:
### Description

In a cluster with a large number of streaming tasks (~1000), `SegmentAllocateAction`s on the overlord can often take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in lag building up while a task waits for a segment to get allocated.

The root causes are:
- large number of metadata calls made to the segments and pending segments tables
- `giant` lock held in `TaskLockbox.tryLock()` to acquire task locks and allocate segments

Since the contention typically arises when several tasks of the same datasource try to allocate segments for the same interval/granularity, the allocation run times can be improved by batching the requests together.

### Changes

#### Broad strokes
- Add flags
   - `druid.indexer.tasklock.batchSegmentAllocation` (default `false`)
   - `druid.indexer.tasklock.batchAllocationMaxWaitTime` (in millis) (default `1000`)
- Put a bound on the amount of time an allocation action can take by configuring the `batchAllocationMaxWaitTime`
- Add `SegmentAllocationQueue` to queue and batch allocate actions
- Identify a batch uniquely with these fields
  - `dataSource`
  - `groupId`
  - `preferredAllocationInterval`
  - `skipSegmentLineageCheck`
  - `useNonRootGenerationPartitionSpace` (as a proxy for `PartialShardSpec.getClass()`)
  - `lockGranularity`
- Metadata calls are now batched
- `giant` lock is acquired just once per batch per tryInterval in `TaskLockbox`
- Rest of the behaviour (order of steps, retries, etc.) in the new flow remains exactly the same as the old flow

#### Adding a request
- Add methods `canPerformAsync` and `performAsync` to `TaskAction`
- `performAsync` currently throws an `UnsupportedOperationException` for other action types
- Submit each allocate action to a `SegmentAllocationQueue`
- Add the action to a batch and execute it when it is due

#### Processing a request batch
- Wait for `maxWaitTimeMillis` before processing a batch
- Send all requests in the batch to `TaskLockbox` to acquire task locks and allocate segments
- Respond to leadership changes and fail items in queue when not leader
- Emit batch and request level metrics

#### Allocating segments
- Add `TaskLockbox.allocateSegments()` retaining the logic from `TaskLockbox.tryLock()`
- Add `IndexerMetadataStorageCoordinator.allocatePendingSegments()` retaining the logic from `allocatePendingSegment()`

#### New metrics
Per batch (dims: `dataSource`, `taskActionType=segmentAllocate`):
- `task/action/batch/runTime`
- `task/action/batch/queueTime`
- `task/action/batch/size`
- `task/action/batch/attempts`

Per request (dims: `taskId`, `taskType`, `dataSource`, `taskActionType=segmentAllocate`):
- `task/action/failed/count`
- `task/action/success/count`

### Pending changes
- Docs for metrics
- Perf eval of new flow against old flow

### Further improvements
- An improvement could be to free up the HTTP threads on which the request is received by using
`HttpServletRequest.startAsync`.
- There are some metadata queries which have not been batched such as updating the
acquired task lock in the metadata store and fetching a matching pending when `skipLineageCheck = false`.

#### Release note
Speed up segment allocation and reduce metadata calls by clubbing multiple requests together.
Set `druid.indexer.tasklock.batchSegmentAllocation=true` to enable this feature.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java:
```diff
@@ -76,11 +76,34 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
+  {
+    try {
+      SegmentAllocateAction allocateAction = (SegmentAllocateAction) taskAction;
+      return (R) allocateAction.performAsync(task, toolbox).get();
```
The `get` call should have a timeout. Even a 10 minute timeout is fine but it shouldn't block forever.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java:
```diff
@@ -76,11 +76,34 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
+  private <R> boolean isBatchAllocateAction(TaskAction<R> taskAction)
+  {
+    return toolbox.canBatchSegmentAllocation()
+        && taskAction instanceof SegmentAllocateAction;
```
could we instead add a method in TaskAction class itself? 
`shouldPerformAsync` and then call `performAsync` instead? 
The method can take toolbox as a parameter and thus the impl in `SegmentAllocationAction` can also check `toolbox.canBatchSegmentAllocation()`

kfaraz replied:
Add `canPerformAsync` and `performAsync` to `TaskAction`.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+  @LifecycleStart
+  public void start()
+  {
+    log.info("Starting queue.");
```
```suggestion
    log.info("Initializing segment allocation queue.");
```

kfaraz replied:
Done.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+  @LifecycleStop
+  public void stop()
+  {
+    log.info("Stopping queue.");
```
```suggestion
    log.info("Tearing down queue.");
```

kfaraz replied:
Done.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+      log.info("Elected leader. Starting queue processing.");
+      scheduleQueuePoll(maxWaitTimeMillis);
+    } else {
+      log.info("Elected leader but batched segment allocation is disabled.");
```
```suggestion
      log.info("Elected leader but batched segment allocation is disabled. segment allocation queue will not be used.");
```

kfaraz replied:
updated.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+
+  public void stopBeingLeader()
+  {
+    log.info("Not leader anymore. Stopping queue processing.");
```
should have different logging depending on whether the batch allocation is enabled or not.

abhishekagarwal87 replied:
you could have also added a no-op implementation that throws an exception if used.

kfaraz replied:
Do you think a no-op impl would be useful for tests?

kfaraz replied:
Hmm, I think having a no-op implementation would require a factory that creates the appropriate `SegmentAllocationQueue` based on the given configs. But that would be cleaner, I suppose.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+
+    AllocateRequestBatch(AllocateRequestKey key)
+    {
+      log.info("Creating a new batch with key: %s", key);
```
how much logging does it result in?

kfaraz replied:
Shouldn't be a lot. If there are too many tasks running, chances are they would end up sharing a batch.
This log information could be useful when not looking at metrics and trying to verify if batches are getting added to the queue. But I guess it can be a debug log line.

Maybe I should just add a log line in the processing method which prints the current queue size.

kfaraz replied:
Updated these logs to be less verbose.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    synchronized void merge(AllocateRequestBatch batch)
+    {
+      requestToFuture.putAll(batch.requestToFuture);
+      batch.requestToFuture.clear();
```
`transferFrom` seems a better name.

kfaraz replied:
Thanks for the suggestion!

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    /**
+     * This must be accessed through methods synchronized on this batch.
+     * It is to avoid races between a new request being added just when the batch
+     * is being processed.
```
can you add few lines about what future objects in map are for?

kfaraz replied:
added.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    {
+      if (!requestToFuture.isEmpty()) {
+        log.info("Marking [%d] requests in batch [%s] as failed.", size(), key);
+        requestToFuture.values().forEach(future -> future.complete(null));
```
shouldn't this be future.completeExceptionally?

abhishekagarwal87 replied:
invoking future calls can also result in an exception. Ideally, the 4 lines below should be done first. Or you can wrap the `future.complete` in a try/catch block.

kfaraz replied:
> shouldn't this be future.completeExceptionally?

No, this is in order to retain the existing behaviour.

kfaraz replied:
> invoking future calls can also result in an exception. Ideally, the 4 lines below should be done first. Or you can wrap the future.complete in a try/catch block.

Oh, I didn't realize. Are you sure? I am just completing the future here and setting a new value in it. 
The only buggy scenario I can think is trying to complete an already completed future, but it doesn't throw an exception even then.

abhishekagarwal87 replied:
I meant that there can be an exception in a listener attached to this future. e.g. 
```
future.whenComplete() {
  throw some exception; 
}
```

But if the callback is a simple op, its fine.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+        return;
+      }
+
+      log.info("Failed to allocate segment for action [%s]: %s", request.getAction(), result.getErrorMessage());
```
```suggestion
      log.warn("Failed to allocate segment for action [%s]: %s", request.getAction(), result.getErrorMessage());
```

abhishekagarwal87 replied:
can this action be anything else other than SegmentAllocationAction?

kfaraz replied:
No, this queue is only for batching `SegmentAllocateAction`s only.

kfaraz replied:
I think I will just remove this log line and include the error message in the subsequent logs inside the if-else.

kfaraz replied:
Updated.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    private final boolean useNonRootGenPartitionSpace;
+
+    private final int hash;
+    private final String serialized;
```
does this string need to be stored? Have you thought about possible GC impact of storing these requests as batches? what will happen if segment allocation is slow for some reason? would Overlord continue to take in requests?

kfaraz replied:
> does this string need to be stored?

No, I think we can get rid of this. Not sure why I had done this in the first place. The computations in a batch are heavy enough to not make this matter at all. 😅

kfaraz replied:
> Have you thought about possible GC impact of storing these requests as batches?

Thanks for the reminder, I do need to do some basic GC evaluation.

kfaraz replied:
> what will happen if segment allocation is slow for some reason? would Overlord continue to take in requests?

Currently, yes. We can try to put a limit on the queue size and/or total number of requests in queue. What do you think?

abhishekagarwal87 replied:
yeah. we don't want the Overlord to crash with an OOM. So having some guardrail will be better.

kfaraz replied:
Put a limit of 5000 on the queue size.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+      if (computedBatch == null) {
+        computedBatch = new AllocateRequestBatch(key);
+        computedBatch.resetQueueTime();
+        processingQueue.offer(computedBatch);
```
what if `offer` returns `false`?

kfaraz replied:
Thanks, will handle this. If the batch couldn't be added to the queue (due to the queue being full), I suppose we don't have a choice but to just complete the request with a `null` return value.

kfaraz replied:
Updated.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    keyToBatch.compute(batch.key, (key, existingBatch) -> {
+      if (existingBatch == null) {
+        batch.resetQueueTime();
+        processingQueue.offer(batch);
```
result checking?

kfaraz replied:
Thanks, will update this.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+  {
+    // If not leader, clear the queue and do not schedule any more rounds of processing
+    if (!leaderSelector.isLeader()) {
+      log.info("Not leader anymore. Clearing [%d] batches from queue.", processingQueue.size());
+      processingQueue.clear();
+      keyToBatch.clear();
```
shouldn't this process take care of the backlog instead of just clearing it?

kfaraz replied:
Yeah, that makes sense. We can complete the pending futures here with a `null` so that the tasks can retry with the new leader.

kfaraz replied:
Updated.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+        ++numProcessedBatches;
+      }
+      catch (Throwable t) {
+        processed = true;
```
this should be false, right? so that we can requeue it.

kfaraz replied:
I am not too sure about this one. An exception being caught here means that there is probably some bug in the code. Retrying wouldn't help. We could maybe raise an alert here? Let me know what you think.

abhishekagarwal87 replied:
well there could be db failures for example. If we don't retry, we should surface the failure to the caller.

kfaraz replied:
We can complete the futures exceptionally. That would also be in line with what the current code would do in case of an exception. Thanks for pointing this out.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+
+    AllocateRequestBatch nextBatch = processingQueue.peek();
+    while (nextBatch != null && nextBatch.isDue()) {
+      processingQueue.poll();
```
what if another batch gets added to processingQueue in this time after line 197?

kfaraz replied:
A new batch may get added but that shouldn't be a problem as only thread would be polling from the queue at any given time. So the result returned by a `poll` would be the same as the result returned by the preceding `peek`.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    int numProcessedBatches = 0;
+
+    AllocateRequestBatch nextBatch = processingQueue.peek();
+    while (nextBatch != null && nextBatch.isDue()) {
```
what if the head of the queue is not due but others in queue are?

kfaraz replied:
That cannot happen, the batch due time is nothing but queuedAtTime + maxWaitTime.
So the batch at the head of the queue has to be the one that will be processed first.
When requeueing a batch, it's queue time is reset.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+    }
+
+    // Key was not found, returned index is (-(insertionPoint) - 1)
+    index = -(index + 1);
```
where is index returned?

kfaraz replied:
`index` is returned from the binary search. I will rephrase the comment.

kfaraz replied:
updated.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,685 @@
+
+    // If the interval at index doesn't overlap, (index + 1) wouldn't overlap either
+    if (index < sortedIntervals.length) {
+      if (sortedIntervals[index].overlaps(searchInterval)) {
```
index can be negative too.

kfaraz replied:
Handled in line 366. I think a test for this method would be helpful.

kfaraz reviewed (commented):
Thanks for the thorough feedback, @abhishekagarwal87 ! I will address your comments and add some more tests to verify the batching behaviour.

rohangarg commented:
One note : While I thought that batching was probably the way to go earlier, probably it could be re-considered in favour of caching and combining it with more parallelism if that's easier. The current batching idea also caches data in memory for pendingSegments table and updates the in-memory state for all requests in a batch to resolve their partition-num assignment conflict.
In a local experiment, I was trying the following : 
1. Add striped locking for `(datasource, interval, sequence name)` tuple in `allocatePendingSegment` method
2. The `createNewSegment` method is the critical section which requires a global lock, but it operates on cached results of `pendingSegments` table. The cache could be for key `(datasource, interval)` and gets invalidated with time, access-count or some transaction failure while committing to pendingSegments table
3. The write to pendingSegments table is also done in the striped lock
4. More things could be extracted from `createNewSegment` (like fetching used segments) to the striped locking flow so that there's more parallelism
Generally, the idea was to do DB calls in parallel threads and use cached objects in the global lock.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -777,6 +994,202 @@ private void insertPendingSegmentIntoMetastore(
+    return createdSegments;
+  }
+
+  private SegmentIdWithShardSpec createNewSegment(
```
Would it be possible to utilize the older method `createNewSegment` and pass the number of segments to be created as an argument?
If all shardSpecs for common properties can be created with increasing partitionIds in the same call, it may make things simpler and help with performance as well

kfaraz replied:
I suppose we could pull out the computation of the `overallMaxId` as that remains the same across the requests in a batch (except including the id of the ids allocated so far). The rest of the stuff in this method will be complicated to pull out because each request can have a different version, depending on the underlying lock that it has created/acquired. Plus, it is not computationally heavy and will not give any perf benefit on commoning out either.

To deduplicate code, the older `createNewSegment` can call this method instead.

kfaraz replied:
With time chunk locking, yes. With segment locking, they may be in a batch but still have different versions. This is just because of the way a segment lock is acquired (see `TaskLockbox.tryLock()` and `LockRequestForNewSegment.getVersion()`).  The datasource, groupId and interval are always the same in both the cases. (see `AllocateRequestKey` in `SegmentAllocationQueue`).

The new code is written like this in order to minimize risk by retaining the same behaviour as the old flow except club the metadata calls together.

AmatyaAvadhanula replied:
Thanks for the clarification

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -183,19 +205,25 @@ private void processBatchesDue()
+
+      AllocateRequestBatch nextBatch = processingQueue.pollFirst();
+      while (nextBatch != null) {
+        nextBatch.markCompleted();
```
Note: Need to make sure that this doesn't keep failing batches and meanwhile the overlord gets re-elected as the leader. Probably re-check the leader condition in the while loop.

kfaraz replied:
Done.

AmatyaAvadhanula commented on core/src/main/java/org/apache/druid/java/util/common/Intervals.java:
```diff
@@ -68,6 +72,50 @@ public static boolean isEternity(final Interval interval)
+    // Key was not found, index returned from binarySearch is (-(insertionPoint) - 1)
+    index = -(index + 1);
+
+    // If the interval at (index - 1) doesn't overlap, (index - 2) wouldn't overlap either
```
This seems like an incorrect assumption with segments of different granularities.

Consider the sorted intervals [ (2022-01-01 / 2023-01-01), (2022-05-01 / 2022-06-01) ]
If we were to take target interval [2022-07-01 / 2022-08-01], index = 1 would not overlap but index = 0 would.

If the intervals to this method are all of the same granularity, could you please add it to the javadoc?

kfaraz replied:
Thanks for catching this, @AmatyaAvadhanula ! Let me fix up the method.

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,694 @@
+      return true;
+    }
+
+    // Requeue the batch only if used segments have changed
```
Could you please explain a bit more about this?

kfaraz replied:
This is to retain behaviour from the existing flow. If the set of used segments has changed since we started processing the request, it's possible that retrying might make it pass.
https://github.com/apache/druid/blob/79df11c16cd2fd1c6a9867855c42768a079917ff/indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocateAction.java#L223

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -446,6 +449,140 @@ public LockResult tryLock(final Task task, final LockRequest request)
+        holderList.getPending().forEach(holder -> acquireTaskLock(holder, false));
+      }
+
+      // TODO: for failed allocations, cleanup newly created locks from the posse map
```
Does this need to be addressed as part of these changes?

kfaraz replied:
Yeah, let me check this. If not done, this might leave some garbage locks in the posse map.

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,694 @@
+      log.debug("Added a new batch [%s] to queue.", batch.key);
+      return true;
+    } else {
+      log.warn("Cannot add batch [%s] as queue is full. Failing [%d] requests.", batch.key, batch.size());
```
what happens as a result of this? I assume it will be caused when the metadata store is slow. can there be any other reason? should we add some possible corrective action here like checking metadata store sizing?

kfaraz replied:
It can also happen when things keep getting retried, due to frequent updates to the set of used segments. It would be up to the client to retry in such cases.

The action can be to look at the emitted metrics `batch/runTime` and `batch/attempts`. A greater value of the first would be caused by metadata slowness, the second is retries. The user can then determine if the metadata store needs resizing. What do you think?

abhishekagarwal87 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,694 @@
+    }
+
+    // Check once again for leadership
+    if (!isLeader.get()) {
+      return;
+    }
```
is this necessary? what if some items were added to the queue just before line 225?

kfaraz replied:
I guess that is not likely to happen. New items are rejected if we are not the leader.
But I guess it could happen if the leader is flapping?
- t1: we are not leader, we clear the queue
- t2: become leader
- t3: add a request
- t4: stop being leader
- t1: we are not leader, return
the request added by t3 will thus remain stuck in the queue, until we become leader again and process it. The future will timeout after 5 minutes.

I guess a simpler alternative to all of this can be to have the leadership check done for each batch separately in `processBatch`. If not leader, fail it there. The only difference would be that even non-leaders would keep polling the queue. I wanted to avoid that somehow. Please let me know what you think.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -777,6 +994,202 @@ private void insertPendingSegmentIntoMetastore(
+
+    for (SegmentCreateRequest request : requests) {
+      // Check if the required segment has already been created in this batch
+      final String sequenceHash = getSequenceNameAndPrevIdSha(request, interval, skipSegmentLineageCheck);
```
When can this occur?
Also, SegmentCreateRequest states that requests must be treated distinctly even if all fields are the same.

kfaraz replied:
Multiple requests for the same segment, in the case of task replicas, I should think.

kfaraz replied:
The `SegmentCreateRequest` is different, the segment id created is not different.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:
```diff
@@ -777,6 +994,202 @@ private void insertPendingSegmentIntoMetastore(
+    // If there is an existing chunk, find the max id with the same version as the existing chunk.
+    // There may still be a pending segment with a higher version (but no corresponding used segments)
+    // which may generate a clash with an existing segment once the new id is generated
+    final SegmentIdWithShardSpec overallMaxId =
```
The overallMaxId is computed assuming that committedMaxId has already added to the set of pendingSegments.
I think it's better to not make that assumption and add committedMaxId explicitly in this method

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java:
```diff
@@ -446,6 +449,140 @@ public LockResult tryLock(final Task task, final LockRequest request)
+        holderList.getPending().forEach(holder -> acquireTaskLock(holder, false));
+      }
+
+      // TODO: for failed allocations, cleanup newly created locks from the posse map
```
I need to verify if this is an actual issue. Also, this behaviour is the same in the old flow too.
I think I should tackle this in a follow up PR, so that this PR just adds the new flow without modifying any behaviour.

kfaraz commented:
@AmatyaAvadhanula , @abhishekagarwal87 , thanks for the review. I have addressed your feedback.

Changes made:
- Queue polling is now always active irrespective of being leader or not. This can be improved later if needed.
- Queue polling starts on lifecycle start and not on becoming leader.
- A future now either completes successfully or exceptionally, i.e. it never completes with null. This allows propagation of actual error message to the caller. Added tests for this.
- Clarify usage of `committedMaxId` in the set of `pendingSegments`
- Rename metric `batch/retries` to `batch/attempts`

Changes in follow up PR:
- docs (mostly just metrics)
- cleanup of empty locks from posse map (both old and new flow)
- limit number of requests in each batch
- cancel the future if it times out on `LocalTaskActionClient`

AmatyaAvadhanula commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -0,0 +1,711 @@
+      final Map<Interval, List<SegmentAllocateRequest>> overlapIntervalToRequests = new HashMap<>();
+
+      for (SegmentAllocateRequest request : allRequests) {
+        // If there is an overlapping used segment, the interval of the used segment
```
The allocation for a single segment considers overlap with queryGranularity.bucket(row.timestamp())
This code fetches all used segments overlapping with segmentGranularity.bucket(row.timestamp()). I think this could lead to a case where the row is allocated to a used segment interval to which it doesn't belong

kfaraz replied:
The used segments for the allocate interval (`segmentGranularity.bucket(row.timestamp())`) are fetched so that we don't have to make the metadata call multiple times.
The logic hasn't changed because we use the `rowInterval` (`queryGranularity.bucket(row.timestamp())`) to find the final overlapping segments. Please look at the line after this.

kfaraz commented:
All unit and integration tests have run successfully with segment allocation enabled in test PR #13484 .
There was only one flaky test. I couldn't retry that test due to some Travis access issues.

kfaraz commented:
Merging as build is green. Latest commit is only comment change.
